lib/vdsm/virt/thinp.py

```diff
@@ -211,83 +215,108 @@ def on_block_threshold(self, target, path, threshold, excess):
                 'Unknown drive %r for vm %s - ignored block threshold event',
                 drive_name, self._vm.id)
         else:
-            drive.on_block_threshold(path)
+            if drive.on_block_threshold(path):
+                periodic.VolumeWatermarkMonitor.dispatch(self._vm)
```

Comment:
How is it ensured that two extensions cannot run concurrently?

Reply:
Concurrent watermark monitoring can be scheduled at the same time, for example:
- The monitor wakes up at the same time an event is handled
- Receiving multiple events (for different drives at the same time)

Multiple threads monitoring at the same time should be safe, since this class does not manage its state but drives state, and drives state is protected by the drive lock.
But this can lead to duplicate extend messages and confusing logs, so I would like to avoid this.

It will be hard to synchronize the dispatch of the periodic watermark monitor and the event based handlers, so if we want to ensure that only single thread is monitoring at the same time, we better ensure it at monitor_volumes.

I think that taking a lock in monitor_volumes using a 2.0 seconds timeout will avoid the issue without risking deadlocks.

Reply:
Sounds reasonably.  But also related to the previous comment — there is a question what to do if the timeout occurs, either in the periodic operation or in monitor_volumes.

Reply:
On timeout the thread just skip this cycle, and we will check on the next cycle, like any other periodic monitior, no?

Reply:
So letting the timed out thread to finish its job and proceeding with new checks, as explained above.  OK.

Reply:
So letting the timed out thread to finish its job and proceeding with new checks, as explained above.  OK.